eth/handler_bor.go

```diff
@@ -39,80 +39,148 @@ var (
+		return blockNum, blockHash, errMilestone
+	}
+
+	// check if we have the checkpoint blocks
+	head := h.ethAPI.BlockNumber()
+	if head < hexutil.Uint64(checkpoint.EndBlock.Uint64()) {
+		log.Debug("Head block behind checkpoint block", "head", head, "milestone end block", checkpoint.EndBlock)
+		return blockNum, blockHash, errMissingMilestone
+	}
+
+	// verify the root hash of checkpoint
+	roothash, err := h.ethAPI.GetRootHash(ctx, checkpoint.StartBlock.Uint64(), checkpoint.EndBlock.Uint64())
+	if err != nil {
+		log.Debug("Failed to get root hash of checkpoint while whitelisting", "err", err)
+		return blockNum, blockHash, errRootHash
+	}
+
+	if roothash != checkpoint.RootHash.String()[2:] {
+		log.Warn("Checkpoint root hash mismatch while whitelisting", "expected", checkpoint.RootHash.String()[2:], "got", roothash)
+		return blockNum, blockHash, errCheckpointRootHashMismatch
+	}
+
+	// fetch the end checkpoint block hash
+	block, err := h.ethAPI.GetBlockByNumber(ctx, rpc.BlockNumber(checkpoint.EndBlock.Uint64()), false)
+	if err != nil {
+		log.Debug("Failed to get end block hash of checkpoint while whitelisting", "err", err)
+		return blockNum, blockHash, errEndBlock
+	}
+
+	hash := fmt.Sprintf("%v", block["hash"])
+
+	blockNum = checkpoint.EndBlock.Uint64()
+	blockHash = common.HexToHash(hash)
+
+	return blockNum, blockHash, nil
+}
+
+// fetchWhitelistMilestones fetches the latest milestone/s from it's local heimdall
+// and verifies the data against bor data.
+func (h *ethHandler) fetchWhitelistMilestone(ctx context.Context, bor *bor.Bor, first bool) (uint64, common.Hash, error) {
+	// Create an array for block number and block hashes
+	//nolint:prealloc
+	var (
+		blockNum  uint64
+		blockHash common.Hash
+	)
+
+	// Fetch the milestone count from heimdall
+	count, err := bor.HeimdallClient.FetchMilestoneCount(ctx)
+	if err != nil {
+		log.Debug("Failed to fetch count for milestones", "err", err)
+		return blockNum, blockHash, errMilestoneCount
 	}
 
 	if count == 0 {
-		return blockNums, blockHashes, errNoCheckpoint
+		return blockNum, blockHash, errNoMilestone
+	}
+
+	//fetch latest milestone
+	milestone, err := bor.HeimdallClient.FetchMilestone(ctx)
+	if err != nil {
+		log.Debug("Failed to fetch latest milestone for whitelisting", "err", err)
+		return blockNum, blockHash, errMilestone
+	}
+
+	// check if we have the milestone blocks
+	head := h.ethAPI.BlockNumber()
+	if head < hexutil.Uint64(milestone.EndBlock.Uint64()) {
+		log.Debug("Head block behind milestone block", "head", head, "milestone end block", milestone.EndBlock)
+		return blockNum, blockHash, errMissingMilestone
 	}
 
-	// If we're in the first iteration, we'll fetch last 10 checkpoints, else only the latest one
-	iterations := 1
-	if first {
-		iterations = 10
+	// verify the root hash of checkpoint
+	roothash, err := h.ethAPI.GetRootHash(ctx, milestone.StartBlock.Uint64(), milestone.EndBlock.Uint64())
```

Comment:
it looks like code duplication, doesn't it?

Reply:
Oh I forgot to change the comment. Will do it. Actually we can make the single function also but would need to give sort of flag to it. What will be best way- To keep the single function with flag or Two seperate functions?

Reply:
Could retrying be  a separate function? I'd like to have a general solution without flags